docs/RFCs/package-toml.md

```diff
@@ -0,0 +1,166 @@
+# Manifest TOML format
+
+A proposal for a more formal TOML definition for entropic
+
+- Version: DRAFT
+- Date: 06/06/2019
+- Authors
+  - Todd Kennedy <todd@selfassembled.org>
+
+## Abstract
+
+This document is meant to describe the package manifest format for Entropic,
+a federated solution to package management. Entropic needs to have its own
+manifest file due to conflicts arising from the namespacing entropic uses.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Copyright
+
+This document is copyright by its individual contributors, and available for
+re-use under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
+
+## Introduction
+
+Entropic cannot share a manifest format with the traditional `package.json` format
+associated with the ECMAScript ecosystem for a number of reasons:
+
+1. Entropic is intended to allow for federated package management for more than just the ECMAScript ecosystem
+1. The `package.json` name specifier does not allow for the namespacing that entropic provides
+1. There are recognized deficiencies with using JSON as human-readable metadata format
+
+The project has chosen to standardize on TOML as a format due to various short-comings in the JSON file specification, especially concerns around:
+
+1. Lack of comments
+1. Lack of multiline string support
+1. Designed for ease of generation by machine, not by hand
+
+## Filename
+
+The filename of the manifest **MUST** be `[TK].toml` and it **MUST** live in
+the root directory of the package's filesystem. The root of the package is defined
+as the closest parent directory that contains the manifest file, starting with
+the current directory.
+
+<details>
+<summary>If file name is case-sensitive</summary>
+On case-insensitive filesystems, it **MAY** be presented as `[TK].toml` but care
+should be taken to preserve case when possible.
+</details>
+
+## Format
+
+The format of this file is **REQUIRED** to be in [TOML](https://github.com/toml-lang/toml)
+version 0.5.0 or later as defined by the project.
+
+### Top-Level Data
+
+The manifest **MUST** contain the following top-level keys, in order to be published:
+
+- `"name"` - the canonical name of the package, as a string, including its namespace. e.g. `"toddself@static-pkg.dev/packge2toml"`
```

Comment:
is the idea that the canonical name of imported packages will also include the namespace?

e.g. `import 'toddself@static-pkg.dev/packge2toml'`

Reply:
IIRC yes, but @chrisdickinson or @ceejbot might have a better answer for this.

Reply:
Yes. Additionally, I was thinking about making packages available by the "smallest unique" name at each level too – so if there's no other `package2toml` at a given level, you could require it via `require('package2toml')`. However, that comes with a cost – so we deferred that behavior until later.

Reply:
requiring using the fully qualified name feels like a big UX fail, imo. I think there can/should be some rule enforcement that makes it the default -- see also Mastodon's `@foo` being context-aware, so you don't have to type out someone's fully-qualified name pretty much -ever-.

Reply:
Maybe it's possible for us to only require the package name by symlink'ing the downloaded package into the`node_modules` folder. 

So, `ds` would install `toddself@static-pkg.dev/packge2toml` as `package2toml`. I think this becomes more complex and simply importing the fully qualified name seems less error prone.

Reply:
@olingern aliases solve this use case nicely

Reply:
@zkat:

> requiring using the fully qualified name feels like a big UX fail, imo. I think there can/should be some rule enforcement that makes it the default -- see also Mastodon's `@foo` being context-aware, so you don't have to type out someone's fully-qualified name pretty much -ever-.

Yep, agreed – I want this feature (and aliases!) At the time, I deferred implementing the context-aware requirement until later because when I went to implement it, it seemed like it would have performance and deduping consquences that I wasn't prepared to tackle at the time. (Also: interactions with how legacy packages are unpacked vs. entropic packages.)

My goal was to produce a working system from which we could iterate. It seemed at the time like we could defer aliasing and context-aware import support until later, while reserving the right to re-define our existing dependency item format as a shorthand for an expanded one that supports those features down the line. (I was starting to get the "I'm too far out on a limb without supports" feeling from speccing aliasing and context-awareness without working from a backing implementation, if that makes sense?)

Reply:
makes total sense to me :)